src/Workspaces/Core/Portable/SemanticModelReuse/SemanticModelWorkspaceServiceFactory.SemanticModelWorkspaceService.cs

```diff
@@ -0,0 +1,175 @@
+                var topLevelSemanticVersion = await document.Project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
+                // If we are able to reuse a semantic model, then ensure that this is now the semantic model we're now
+                // pointing at for this document.
+                var reuseInfo = await TryReuseCachedSemanticModelAsync(
+                    map, document, bodyNode, linkedIds, topLevelSemanticVersion, cancellationToken).ConfigureAwait(false);
+                if (reuseInfo != null)
+                    return map.SetItem(document.Id, reuseInfo.Value);
+
+                // Otherwise, we couldn't reuse anything from the cache.  Return a fresh map with just the real semantic
+                // model value in it.
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                var builder = ImmutableDictionary.CreateBuilder<DocumentId, SemanticModelReuseInfo?>();
+
+                // Note: we are intentionally storing the semanticmodel instance in the speculative location as well.
+                builder.Add(document.Id, new SemanticModelReuseInfo(semanticModel, semanticModel, bodyNode, topLevelSemanticVersion));
+
+                foreach (var linkedId in linkedIds)
+                    builder.Add(linkedId, null);
+
+                return builder.ToImmutable();
+            }
+
+            private static async Task<SemanticModelReuseInfo?> TryReuseCachedSemanticModelAsync(
+                ImmutableDictionary<DocumentId, SemanticModelReuseInfo?> map,
+                Document document,
+                SyntaxNode bodyNode,
+                ImmutableArray<DocumentId> linkedIds,
+                VersionStamp topLevelSemanticVersion,
+                CancellationToken cancellationToken)
+            {
+                // Get the clique (i.e. all the ids for this and the other documents it is linked to) corresponds to
+                // matches the clique we're caching.
+                using var _ = PooledHashSet<DocumentId>.GetInstance(out var documentIdClique);
+                documentIdClique.Add(document.Id);
+                documentIdClique.AddRange(linkedIds);
+
+                // if this is asking about a different doc or clique, we can't reuse anything.
+                if (!map.Keys.SetEquals(documentIdClique))
+                    return null;
+
+                // see if this doc matches the docs we're caching information for.
+                if (!map.TryGetValue(document.Id, out var reuseInfoOpt) || !reuseInfoOpt.HasValue)
+                    return null;
+
+                var reuseInfo = reuseInfoOpt.Value;
+
+                // can only reuse the cache if nothing top level changed.
+                if (reuseInfo.TopLevelSementicVersion != topLevelSemanticVersion)
+                    return null;
```

Comment:
Any concerns about whether we're being too eager here if there's preprocessor stuff involved? If you have an #ifdef and one side is in a method body and the other side is the preprocessor trivia, the editing the preprocessor trivia is a top-level edit for that side, but the other side can totally reuse it's semantic model -- and that's the one actually being used by completion.

My imagination for the overall workflow here would have been:

1. First check for DocumentId being in the map, if not generate a new map like you do with the one entry and everything else null.
2. For any other update, you're only _ever_ updating the single entry in the map to avoid disturbing other things.

Reply:
we should chat tomorrow, i don't quite get what you're saying.  "TopLevelSemanticVersion" is only impacted by pp changes if thoes pp changes actuallycaused a top-level tree change.  If it didn't, then you'll have the same version as before.

Reply:
understood now.  now we do keep the other linked informatino if available.